### ISSUE TITLE
Strange behavior while lowering nn.BatchNorm2d

### ISSUE THREAD

**Issue by JBloodless** [nod-ai/SHARK-Turbine#110]
Hello. I'm using the most basic BatchNorm in my model and for some reason it is being lowered to some strange variation, which is not supported.

Here's minimal repro:

```
import shark_turbine.aot as aot
import torch
import torch.nn as nn

class AdaptCNN(nn.Module):
    def __init__(self,
                 input_channels=1,
                 c_out_1=16,
                 kernel_size=(3, 3),
                 ):
        super().__init__()

        self.input_channels = input_channels
        self.c_out_1 = c_out_1
        self.kernel_size = kernel_size

        self.cnn_pad = (1, 1)

        self.conv1 = nn.Conv2d(
            self.input_channels,
            self.c_out_1,
            self.kernel_size,
            padding=self.cnn_pad)

        self.bn1 = nn.BatchNorm2d(self.conv1.out_channels)

    def forward(self, x):
        x = self.bn1(self.conv1(x))
        return x


x = torch.zeros((63, 1, 48, 15))

model = AdaptCNN()
model.eval()

export_output = aot.export(model, x)
binary = export_output.compile(save_to=None)
```

And here's what I get:

```
Traceback (most recent call last):
  File "/Users/i.beskrovnyy/tts/NISQA-s/repro_bn.py", line 38, in <module>
    export_output = aot.export(model, x)
                    ^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/exporter.py", line 198, in export
    cm = Exported(context=context, import_to="import")
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/compiled_module.py", line 534, in __new__
    do_export(proc_def)
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/compiled_module.py", line 531, in do_export
    trace.trace_py_func(invoke_with_self)
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/support/procedural/tracer.py", line 120, in trace_py_func
    return_py_value = _unproxy(py_f(*self.proxy_posargs, **self.proxy_kwargs))
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/compiled_module.py", line 512, in invoke_with_self
    return proc_def.callable(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/exporter.py", line 182, in main
    return jittable(mdl.forward)(*args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/support/procedural/base.py", line 137, in __call__
    return current_ir_trace().handle_call(self, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/support/procedural/tracer.py", line 136, in handle_call
    return target.resolve_call(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/aot/builtins/jittable.py", line 239, in resolve_call
    fx_importer.import_stateless_graph(gm.graph, func_name=self.function_name)
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/dynamo/importer.py", line 265, in import_stateless_graph
    node_importer.import_nodes(g.nodes)
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/dynamo/importer.py", line 496, in import_nodes
    self._import_torch_op_overload(loc, node, target)
  File "/Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/shark_turbine/dynamo/importer.py", line 633, in _import_torch_op_overload
    raise NotImplementedError(
NotImplementedError: Unimplemented torch op in the IREE compiler: 'torch.aten._native_batch_norm_legit_no_training' (either implement this op/variant or configure the compiler to allow unknown operations and fallback to PyTorch).
```

Why BatchNorm fallbacks to this `_native_batch_norm_legit_no_training` and what am I doing wrong here?

**Comment by dan-garvey** [comment]
Thanks for filing this! You aren't doing anything wrong. We need to update the pypi version of turbine. If you do `pip uninstall shark-turbine` and then navigate to a clone of shark-turbine, `pip install -e .`, then your code should work without issue. This issue was solved by #102 . We'll push some new pip packages soon. Let me know if this works for you and I can close the issue.

**Comment by JBloodless** [comment]
> Thanks for filing this! You aren't doing anything wrong. We need to update the pypi version of turbine. If you do `pip uninstall shark-turbine` and then navigate to a clone of shark-turbine, `pip install -e .`, then your code should work without issue. This issue was solved by #102 . We'll push some new pip packages soon. Let me know if this works for you and I can close the issue.

Export works with latest turbine (I guess), but `.compile` throws

```
Traceback (most recent call last):
  File "/Users/i.beskrovnyy/tts/NISQA-s/repro_bn.py", line 40, in <module>
    binary = export_output.compile(save_to=None)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/i.beskrovnyy/SHARK-Turbine/python/shark_turbine/aot/exporter.py", line 138, in compile
    raise RuntimeError("Compilation failed: See diagnostics")
RuntimeError: Compilation failed: See diagnostics
```

**Comment by zjgarvey** [comment]
> > Thanks for filing this! You aren't doing anything wrong. We need to update the pypi version of turbine. If you do `pip uninstall shark-turbine` and then navigate to a clone of shark-turbine, `pip install -e .`, then your code should work without issue. This issue was solved by #102 . We'll push some new pip packages soon. Let me know if this works for you and I can close the issue.
> 
> Export works with latest turbine (I guess), but `.compile` throws
> 
> ```
> loc("<eval_with_key>.0 from /Users/i.beskrovnyy/anaconda3/envs/mlir/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py:489 in wrapped":8:0): error: 'tensor.cast' op operand type 'tensor<?xui8>' and result type 'tensor<0xi8>' are cast incompatible
> Traceback (most recent call last):
>   File "/Users/i.beskrovnyy/tts/NISQA-s/repro_bn.py", line 40, in <module>
>     binary = export_output.compile(save_to=None)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/i.beskrovnyy/SHARK-Turbine/python/shark_turbine/aot/exporter.py", line 138, in compile
>     raise RuntimeError("Compilation failed: See diagnostics")
> RuntimeError: Compilation failed: See diagnostics
> ```

Here is a minimal reproduction:



```
import shark_turbine.aot as aot
import torch
import torch.nn as nn

class ExMod(nn.Module):
    def __init__(self):
        super().__init__()
        self.m = nn.BatchNorm2d(100)

    def forward(self,x):
        return self.m(x)

x = torch.zeros(20,100,35,45)

mod = ExMod()
mod.eval()

export_output = aot.export(mod,x)
#export_output.save_mlir('bnex.mlir')
binary = export_output.compile(save_to = None)
```

**Comment by JBloodless** [comment]
Bumping this issue since not having working `nn.BatchNorm2d `looks crucial (it's one of the most basic operations in models)

**Comment by stellaraccident** [comment]
We may be missing patterns to elide zero element tensors.

**Comment by AmosLewis** [comment]
```
(iree_venv) ➜  src  cd /nodclouddata/chi/src ; /usr/bin/env /nodclouddata/chi/s
rc/SHARK-Turbine/.venv/bin/python3.11 /home/chi/.vscode-server/extensions/ms-py
thon.python-2023.20.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/laun
cher 38747 -- /nodclouddata/chi/src/SHARK-Turbine/tests/aot/batchnorm2d_test.py
 
loc("<eval_with_key>.0 from /nodclouddata/chi/src/SHARK-Turbine/.venv/lib/python3.11/site-packages/torch/fx/experimental/proxy_tensor.py:477 in wrapped":5:0): 
error: 'tensor.cast' op operand type 'tensor<?xui8>' and result type 'tensor<0xi8>' are cast incompatible
```
The key debug info is **'tensor.cast' op operand type 'tensor<?xui8>' and result type 'tensor<0xi8>' are cast incompatible**

**Comment by AmosLewis** [comment]
One more bug: the `export_output.save_mlir('bnex.mlir') `with TOM iree will lead to this. But with shark_turbine `iree-compiler=20131113.707` is good.
```
TypeError: print(): incompatible function arguments. The following argument types are supported:
    1. (self: iree.compiler._mlir_libs._mlir.ir._OperationBase, state: mlir::python::PyAsmState, file: object = None, binary: bool = False) -> None
    2. (self: iree.compiler._mlir_libs._mlir.ir._OperationBase, large_elements_limit: Optional[int] = None, enable_debug_info: bool = False, pretty_debug_info: bool = False, print_generic_op_form: bool = False, use_local_scope: bool = False, assume_verified: bool = False, file: object = None, binary: bool = False) -> None

Invoked with: <iree.compiler._mlir_libs._mlir.ir.Operation object at 0x7f9fea26e030>, <_io.BufferedWriter name='bnex.mlir'>; kwargs: binary=True
```

**Comment by JBloodless** [comment]
Is there any progress/workaround yet?

**Comment by AmosLewis** [comment]
> Is there any progress/workaround yet?

Yes. Should be fixed this week.

**Comment by AmosLewis** [comment]
> Is there any progress/workaround yet?
For the upstream fix, we will need to push a fix patch in torch-mlir upstream and then uplift the torch-mlir version IREE, then uplift IREE version shark-turbine use, it will take some time. So here I provide a local quick fix for your use.

For the **runtime error**.
Here is a quick fix locally in IREE thirdparty torch-mlir.
To build iree, you need to look at [IREE getting started](http://iree.dev/building-from-source/getting-started/#prerequisites)
Then once you setup iree build env(including the python env), you need to manually changed code here https://github.com/shark-infra/torch-mlir/blob/ad182198204e299377d4efba74fa2241e5404b0f/lib/Dialect/Torch/IR/TorchTypes.cpp#L405C1-L406C52
Replace the 
```
    return IntegerType::get(context, integerType.getWidth(),
                            IntegerType::Signless);
```
with 
```
    return dtype;
```
Then build iree locally
```
cmake -G Ninja -B ../iree-build/ -S . \
    -DCMAKE_BUILD_TYPE=Release \
    -DIREE_ENABLE_ASSERTIONS=ON \
    -DIREE_ENABLE_SPLIT_DWARF=ON \
    -DIREE_ENABLE_THIN_ARCHIVES=ON \
    -DCMAKE_C_COMPILER_LAUNCHER=ccache \
    -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
    -DIREE_BUILD_PYTHON_BINDINGS=ON \
    -DPython3_EXECUTABLE="$(which python)" \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DIREE_ENABLE_LLD=ON
cmake --build ../iree-build/
```
Then in order to use the new build python bindings, in your shark-turbine python env you have to 
```
pip uninstall iree-compiler
pip uninstall iree-runtime
```
And set PYTHONPATH with the iree new build python bindings, here is an example in my path:
`export PYTHONPATH=/nodclouddata/chi/src/iree-build/compiler/bindings/python:/nodclouddata/chi/src/iree-build/runtime/bindings/python:/nodclouddata/chi/src/SHARK-Turbine/python:$PYTHONPATH`

For the **save_mlir incompatible error**: you can use  `export_output.print_readable()` instead

**Comment by AmosLewis** [comment]
@JBloodless  With the TOM torch-mlir the issue is fixed now. We will next uplift the change into IREE and then shark-turbine.

**Comment by AmosLewis** [comment]
@JBloodless Now you can fix your issue in Shark-Turbine Python env by
```
pip install iree-compiler==20231130.724
pip install iree-runtime==20231130.724
```

**Comment by AmosLewis** [comment]
> > iree-runtime==20231130.724
> 
> ```
> (mlir) i.beskrovnyy@i-beskrovnyy-2 NISQA-s % pip install iree-compiler==20231130.724
> ERROR: Could not find a version that satisfies the requirement iree-compiler==20231130.724 (from versions: 20230404.479, 20230419.494, 20230524.529, 20230815.614, 20231004.665, 20231113.707)
> ERROR: No matching distribution found for iree-compiler==20231130.724
> (mlir) i.beskrovnyy@i-beskrovnyy-2 NISQA-s % pip install iree-runtime==20231130.724
> ERROR: Could not find a version that satisfies the requirement iree-runtime==20231130.724 (from versions: 20230228.444, 20230404.479, 20230419.494, 20230524.529, 20230815.614, 20231004.665, 20231113.707)
> ERROR: No matching distribution found for iree-runtime==20231130.724
> ```
> 
> Seems like these versions are not in pypi yet, but I'll keep monitoring after the weekend, thanks

You need the `-f https://openxla.github.io/iree/pip-release-links.html`. You can find it in the **requirements.txt**. What I usually do is change [requirements.txt](https://github.com/nod-ai/SHARK-Turbine/blob/main/requirements.txt) and `pip install -r requirements.txt`

**Comment by JBloodless** [comment]
Closing this since BatchNorm2d seems to be fixed with `20231130.724`
